Utilities Notebook

olivialynn opened the pull request:
## Change/Solution Description
<!--- 
Describe your changes in detail. In your description, you should answer questions like "Why is this change required? What problem does it solve?".

If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged.
-->
This PR presents a new notebook to gather utilities that mentioned in various rail notebooks (or have so far only been mentioned in live rail demos) to give new and existing users a one stop shop to record and reference rail utilities.

This will also help us reduce clutter in other notebooks, so they may focus purely on the feature they demonstrate.

## Code Quality
- [x] I have read the Contribution Guide
- [x] My code follows the code style of this project
- [x] My code builds (or compiles) cleanly without any errors or warnings
- [x] My code contains relevant comments and necessary documentation

johannct commented on rail/estimation/algos/bpz_lite.py:
```diff
@@ -0,0 +1,376 @@
+                                                  kernel, flux,
+                                                  flux_err, mag_0,
+                                                  self.zgrid)
+
```
ok the for loop obviously will go away, and its proper multi-core/node distribution should be taken care of by the base class logically. For the time being, I would recommend bringing `self.flux_template` to local scope only once (I think calling `self.` repeatedly is an overhead), and extracting the test_data columns in one shot, and then define the loop as `for i,m0,f,fe in enumerate(zip(m0_col,f_col,fe_col):` or something like that

sschmidt23 replied:
The for loop is only looping over the subset passed to the code in each chunk, e.g. 1000 galaxies out of ~21,000 in the example test data, so I don't think it will go away.  For flux_template, I don't think that the speed loss is fairly minor, and not worth making a copy locally, though we can try it and see.  For the other variables, I find this form much more readable, and I think it involves less copying of data, since we're only grabbing one row of the test_data, rather than copying the whole thing over and enumerating those.

johannct replied:
ti should be a view, not a copy. But I am no numpy memory contiguity either. I just learned the hard way to work column wise as much as possible, and loop over rows only when strictly necessary. As a matter of fact the called function could take arrays as argument, rather than floats. This would move the for loop closer to actual computations

sschmidt23 replied:
Ok, I pulled out the two `self.` references in the loop and also changed this to a list comprehension rather than a for loop.

sschmidt23 replied:
Any combo of the loop looked uglier and less comprehensible to me than the original for loop, and the original loop is also closer to the original BPZ code, so I switched it back.  I did store local copies of the two self variables that are called in the loop, which should save a small amount of time.

johannct commented on setup.py:
```diff
@@ -32,6 +36,7 @@
                       'pyarrow',
                       'tables',
                       'astropy',
+                      'DESC_BPZ @ git+https://github.com/LSSTDESC/DESC_BPZ',
```
any chance you can take the time to turn it into a pip installable package?

sschmidt23 replied:
I asked Alex about this, I think the consensus was that not all installable packages need to be on PyPI, this setup pulling from GitHub seems to work, so we may leave it in that form rather than insisting on all packages being on PyPI.

johannct replied:
DESC CO may insist....

sschmidt23 replied:
Ok, but that issue is beyond this PR, we can create a separate issue somewhere, as this same issue is also true for qp as at least one other RAIL dependency.

johannct replied:
yes

johannct reviewed (changes requested):
- I think that `base.yaml` is now broken :  the paths to trainfile and testfile need to go two level up : `../../`
- irrespective of what the structure will eventually be and how we will deal with provenance, the new dirs AB, SED, FILTERS need each a README.md file explaining where the various files come from and what there nominal usage is. Else in the interim this may still drive us crazy.
- 3 files have a conflict to solve.
Thanks a lot for all this! Let's clean the comments fast so that we can approve.

johannct commented on examples/estimation/configs/BPZ_lite.yaml:
```diff
@@ -0,0 +1,17 @@
+  zmin: 0.0
+  zmax: 3.0
+  dz: 0.01
+  columns_file: ./TMPBPZ/test.columns
```
this seems to be needed but is not provided in the current branch

sschmidt23 replied:
Ah, this was renamed in an internal reorganization that I did and I missed it.  Changing the columns file to point to examples/estimation/configs/test_bpz.columns

sschmidt23 commented on examples/estimation/base.yaml:
```diff
@@ -5,4 +5,4 @@ base_config:
   configpath: ./configs
   outpath: ./results
-  output_format: old
+  output_format: qp
```
I flipped this default while doing some testing and forgot to change it back, I'll revert it back to `old` now.

sschmidt23 commented on examples/estimation/configs/simpleNN.yaml:
```diff
@@ -1,7 +1,7 @@
   class_name: simpleNN
   run_name: test_simpleNN
-  width: 0.065
+  width: 0.025
```
I was playing with some tests and didn't mean to commit this change, reverting.

sschmidt23 commented on examples/estimation/configs/test_bpz.columns:
```diff
@@ -0,0 +1,11 @@
+#Filter        columns  cal(AB or Vega,optional)  zp_error  zp_offset
```
comment added

sschmidt23 commented on rail/estimation/algos/bpz_lite.py:
```diff
@@ -0,0 +1,376 @@
+        return flux_templates
+
+    def make_new_ab_file(self, spectrum, filter):
+        from desc_bpz.bpz_tools_py3 import ABflux
```
It's a remnant from when Will Hartley was working on the update, I think.  It does distinguish it from the py2 version that is floating around.  I'm not bothered by it, but I also wouldn't mind removing the _py3 from filenames.

johannct replied:
as you like. Maybe an issue to open for the next round of improvement?

sschmidt23 replied:
I'll create an issue on DESC_BPZ and mention the renaming

sschmidt23 commented on rail/estimation/algos/bpz_lite.py:
```diff
@@ -0,0 +1,376 @@
+with TXPipe and ceci by Joe Zuntz and Sam Schmidt
+for BPZPipe.  This version for RAIL removes a few
+features and concentrates on just predicting the PDF.
+
```
It actually belongs in both, as some of the BPZ code functionality is copied in bpz_lite as well.  I do need to add more specific acknowledgements to DESC_BPZ after I email Dan Coe to ask a couple of questions.

johannct replied:
oh then the acknowledgement and licensing needs to be a bit more formal, then.

sschmidt23 replied:
The original BPZ code has no license file.  I'm about to email Dan Coe about the Madau bug fix in DESC_BPZ, I'll also ask if it's ok that we have messed with the code, and also about posting to PyPI.  Again, this being a modification of a public code with no official license is an uncomfortable area of space to be in.  I think there may be similar issues with other codes, e.g. Delight with similar issues (though I don't know what license info Delight has).

johannct replied:
Yes let's try to clean things up. The advantage with Delight is that Boris is in DESC.

sschmidt23 replied:
I'm waiting on a reply from Dan Coe, but again, this is an issue for DESC_BPZ rather than bpz_lite, so is there anything remaining to be done here?

sschmidt23 commented on rail/estimation/algos/bpz_lite.py:
```diff
@@ -0,0 +1,376 @@
+        if prior_file == 'flat' or prior_file == 'none':
+            P = np.ones(L.shape)
+        else:
+            P = prior(z, mag_0, prior_file, nt, ninterp=0)  # hardcode interp 0
```
BPZ already reads in a prior from a file specified in the header: the code defines prior as an import from a specific file name, so this already functions as if it was `set_prior` rather than just `prior`

johannct replied:
hmm I have a hard time following the code here, especially in the context of an implementation of `inform`.....

sschmidt23 replied:
This is the way the original code was set up, again, the goal here was to do a minimal port of BPZ, and this is how BPZ is written.  For inform, I think it will entail running a custom fit to find the ~14 parameters, and then a special prior function that somehow reads in those parameters.  I haven't worked out exactly how I wanted to do that yet, which is why `inform` remains unimplemented.

sschmidt23 commented on rail/estimation/algos/bpz_lite.py:
```diff
@@ -0,0 +1,376 @@
+        return data
+
+    def estimate_pdf(self, flux_templates, kernel, flux, flux_err, mag_0, z):
+        from desc_bpz.bpz_tools_py3 import p_c_z_t, prior
```
This is left over from some code that Joe Zuntz wrote, but yes, I'll move all imports to the top of the file.

sschmidt23 replied:
Ah, trying to change this reminded me why the imports are structured like this: bpz_tools_py3 requires that an environment variable be set, which happens in the __init__.  If the imports to bpz_tools_py3 are at the top of the file, an exception occurs because the variable has not yet been set.  

So, I reverted the changes and moved the imports back to within the functions in bpz_lite.

johannct replied:
ok..... sounds like hiding a problem though...

sschmidt23 replied:
Yes, I agree, though the way the original code was written this is how things work.  There's a similar issue with the prior, as the name of the .py file that it imports `prior` from is also in the config file.

sschmidt23 commented on rail/estimation/algos/bpz_lite.py:
```diff
@@ -0,0 +1,376 @@
+            os.environ["BPZDATAPATH"] = tmpdatapath
+            self.data_path = tmpdatapath
+        else:  #pragma: no cover
+            self.data_path = datapath
```
added a check

sschmidt23 commented on examples/estimation/configs/BPZ_lite.yaml:
```diff
@@ -0,0 +1,17 @@
+run_params:
```
Should we deal with this separate from this PR, as it applies to every yaml file in configs?

sschmidt23 commented:
I heard back from Dan Coe via email, he wasn't sure of which License to use and CC'd Noel Benitez on the email (though Benitez has apparently left the field).  I'm not sure how long it will take to hear back on the License issue, which is really an issue for DESC_BPZ not RAIL, so I'm going to resolve the conflict and try to merge this in now, and we can worry about the incomplete readthedocs later.

olivialynn committed:
drafts of prospective new notebooks

olivialynn committed:
drafts of prospective new notebooks

olivialynn committed:
First pass of todos

olivialynn committed:
First pass of todos

olivialynn committed:
Most TODOs addressed

olivialynn committed:
Most TODOs addressed

olivialynn committed:
More iteration

olivialynn committed:
More iteration

olivialynn committed:
rmv draft nb

olivialynn committed:
rmv draft nb

aimalz committed:
this got left behind when migrating from Olivia's fork (#68)

It was in source/index_body.rst on https://github.com/OliviaLynn/RAIL so didn't get propagated to index.rst here, but it's the most visible content on RTD so merits updating!

olivialynn committed:
Merge branch 'user/olynn/new-nbs' of https://github.com/LSSTDESC/rail into user/olynn/new-nbs

olivialynn committed:
Update last run successfuly

olivialynn committed:
I cannot spell

olivialynn committed:
explain prefix omission

olivialynn committed:
rmv todo for toc

olivialynn committed:
rmv outputs

eacharles reviewed (approved):
Look great, thanks.
